e2e tests state preparation for url-cache boots

naz opened the pull request:
refs https://github.com/TryGhost/Toolbox/issues/135

naz committed:
Updated Users API e2e tests to use 'posts' fixture

refs https://github.com/TryGhost/Toolbox/issues/135

- This way the test is performed on a "full" set of posts that will become unified within all e2e tests for performance optimization

naz committed:
Hardcoded id's in test fixtures

refs https://github.com/TryGhost/Toolbox/issues/135

- These fixtures needs id's to be able to form a stable urls and resources cache that would be used across all e2e tests when booting without a frontend

naz committed:
Updated urls/resources cache fixtures

refs https://github.com/TryGhost/Toolbox/issues/135

- These fixture had to be updated due to hardcoded ids introduced in the fixtures.json for test
- This allows to boot up a Ghost instance without recalculating urls for most common database states

naz committed:
Introduced "withFrontend" and "withBackend" flags in boot

refs https://github.com/TryGhost/Toolbox/issues/135

- These flags are meant to control initialization of sections of the boot sequence depending on the needs - with or without bakend (API)/frontend (public handlebars site)
- Ideally these flags should not be passed deep into the components, and if the are (like in the web/parent/app case) it's a smell that we need to move things up into the boot process!

naz committed:
Introduced booting from fixture url cache in test environment

refs https://github.com/TryGhost/Toolbox/issues/135

- To be able to reliably start ghost instance without a frontend the process needs access to urls/resources caches
- Storing the configuration in "paths" for now as there's no better place for it untill we are able to mock the content folder in pre-boot

naz committed:
Optimized e2e-api tests using boot with no frontend

refs https://github.com/TryGhost/Toolbox/issues/135

- When running e2e-API test in most cases there's no need to boot Ghost instance with full frontend. This should improve the boot time which should reflect on the speed of running test suites
- The tests where the "forceStart" and "withFrontend" are used together indicate that there's still some work to do to fully separate frontend/backend boot line. The force start is also unnecessary, but was needed to reinitialize all services properly - should be investigated!

naz committed:
Moved startGhost function to local utils

refs https://github.com/TryGhost/Toolbox/issues/135

- Having the proxy in local utils allows to disable frontend gradually in test suites instead of all at once (the latter approach didn't work out as too many other suites still depend on full frontend initialization)

naz committed:
Renamed testUtils.startGhost to localUtils.startGhost

refs https://github.com/TryGhost/Toolbox/issues/135

- Going though local utils allows to have the "withFrontend: false" flag applied only to the e2e-api test suites. This way we can gradually introduce the no-frontend change across all test suites

naz committed:
Changed the test environment to load withFrontend

refs https://github.com/TryGhost/Toolbox/issues/135

- The global default should stay the same as it used to be and we can introduce an override for "withFrontend:false" on casa-by-case or area-by-area bases

naz committed:
Fixed undefined error

refs https://github.com/TryGhost/Toolbox/issues/135

- Not all callers provide the parameter, so an empty object default saves from unecessary "undefined" errors

naz committed:
Extracted a local file cache class for URLs

refs https://github.com/TryGhost/Toolbox/issues/135

- This extracts the file storage knowledge out of the URL Service an allows to have optional features based on the environment - for example turning off writing cache for when running tests

naz committed:
Added writeDisabled flag to URL file cache

refs https://github.com/TryGhost/Toolbox/issues/135

- Allows to turn off overwriting urls/resources JSON file caches on testing environment. This is needed to have predictable state when running multiple test suites that stop the Ghost process and try to persiste URL cache.

naz committed:
Fixed event listener pileup

refs https://github.com/TryGhost/Toolbox/issues/135

- When instantiating new Urls/Resources object in the UrlService's init on every test suite start it was loosing track of past events (softReset wasn't doing the cleanup properly). Not initializing new classes "fixes" the problem partially, by not loosing track of those event listeners. The real fix should be proper even listener cleanup on every soft reset!

naz committed:
Skipped unnecessary frontend initialization

refs https://github.com/TryGhost/Toolbox/issues/135

- This optimization is expected to play a role in more consistent "backend-only" boot where the previous test state might have left over a different theme version which might cause in unwanted URL Services reainitializations.
- What has been happening here is the themes.test.js suite was uploading a theme with a v4 api and when the users api test suite loaded up it switched back to a default v2 theme, which caused routing reinitialization
- The root problem here  is the themese suite is leaving a mess behind so a "restartModeGhostStart" is not really possible anymore - this should be cleaned up separately

naz committed:
Fixed UrlService initialization

refs https://github.com/TryGhost/Toolbox/issues/135

- When there's no instance of cache available the code to retreive it should not execut. Duh!

naz committed:
Changed /^testing/ regexp use to startsWith method

no issue

- `startsWith` method is way easier to read and understand. also, **probably** has better performance comparing to building up a regexp and then matching

naz committed:
Fixed frontend initialization in "mock-express" tests

refs https://github.com/TryGhost/Toolbox/issues/135

- Without sensible defaults the web app was not initializing either the backend nor the frontned parts of the application. Fixed the defaults so the problem doesn't happen again and optimized mock-express-style initialization to only initialize the frontend routing

ErisDS commented on core/boot.js:
```diff
@@ -346,11 +351,14 @@ async function bootGhost() {
+            ghostServer = new GhostServer({url: config.getSiteUrl()});
+            await ghostServer.start(rootApp);
+            bootLogger.log('server started');
+            debug('End: load server + minimal app');
```
Little thing but this debug statement shouldn't be caught by the if statement

ErisDS replied:
A more hairy comment 😬 

IMO mounting the backend onto the app, and booting the server are not the same thing. I wouldn't expect setting backend/withBackend false to result in the server not being created 🤔 I'd expect it to result in a server without any backend paths mounted.

ErisDS commented on core/boot.js:
```diff
@@ -297,7 +302,7 @@ async function initBackgroundServices({config}) {
  * @returns {Promise<object>} ghostServer
  */
-async function bootGhost() {
+async function bootGhost({withBackend = true, withFrontend = true} = {}) {
```
Also a minor comment, but these params could simply be `backend` and `frontend`. With makes it more confusing IMO, not less 😉

naz commented:
This PR still needs few last bits fixing up the test environment. Hoping to merge it on Monday :crossed_fingers:

naz committed:
Renamed withBackend/withFrotned to backend/frontend

refs https://github.com/TryGhost/Toolbox/issues/135

- Shorter name still makes sense and the "with" might be causing confusion understanding what's gonig on during the boot time

naz committed:
Removed no backend optimization in boot

refs https://github.com/TryGhost/Toolbox/issues/135

- The expectation is to only skip monting the backend API routes instead of skipping creation of whole GhostServer

naz committed:
Removed unneeded mocks

refs https://github.com/TryGhost/Toolbox/issues/135

- These mocks were failing to execute but didn't make any difference in the test outcome
